File: checkers/btc_checker.py
import asyncio

async def check_rest(session, address, url):
    try:
        async with session.get(url + address, timeout=10) as r:
            if r.status != 200:
                return None

            data = await r.json()

            if "chain_stats" in data:
                return data["chain_stats"].get("tx_count", 0) > 0

            elif "txs" in data:
                return len(data["txs"]) > 0

            elif "tx_count" in data:
                return data["tx_count"] > 0

    except:
        return None

    return False


async def check_rpc(session, address, url):
    try:
        payload = {
            "jsonrpc": "1.0",
            "id": "scan",
            "method": "getreceivedbyaddress",
            "params": [address]
        }

        async with session.post(url, json=payload, timeout=10) as r:
            if r.status != 200:
                return None

            data = await r.json()

            if isinstance(data.get("result"), (int, float)):
                return data["result"] > 0

    except:
        return None

    return False


class BTCChecker:
    def __init__(self, rpc_manager):
        self.rpc_manager = rpc_manager

    async def check(self, session, address):
        for _ in range(len(self.rpc_manager.rpcs)):
            entry = self.rpc_manager.get()

            if not entry:
                return False

            url = entry["url"]
            mode = entry.get("mode", "RPC")

            if mode == "REST":
                result = await check_rest(session, address, url)
            else:
                result = await check_rpc(session, address, url)

            if result is True:
                return True

        return False

File: checkers/btcChecker.js
const { fetchJson, postJson } = require("../core/http");

async function checkRest(address, url) {
  const data = await fetchJson(`${url}${address}`);
  if (!data) return null;

  if (data.chain_stats) return (data.chain_stats.tx_count || 0) > 0;
  if (Array.isArray(data.txs)) return data.txs.length > 0;
  if (typeof data.tx_count === "number") return data.tx_count > 0;

  return false;
}

async function checkRpc(address, url) {
  const data = await postJson(url, {
    jsonrpc: "1.0",
    id: "scan",
    method: "getreceivedbyaddress",
    params: [address]
  });

  if (!data) return null;
  return typeof data.result === "number" ? data.result > 0 : false;
}

class BTCChecker {
  constructor(rpcManager) {
    this.rpcManager = rpcManager;
  }

  async check(address) {
    for (let i = 0; i < this.rpcManager.rpcs.length; i += 1) {
      const entry = this.rpcManager.get();
      if (!entry) return false;

      const result = entry.mode.toUpperCase() === "REST"
        ? await checkRest(address, entry.url)
        : await checkRpc(address, entry.url);

      if (result === null) {
        this.rpcManager.markFail(entry);
        continue;
      }

      this.rpcManager.markSuccess(entry);
      if (result === true) return true;
    }

    return false;
  }
}

module.exports = { BTCChecker };
